### PR TITLE
Fix segfault when reflection padding out of range

### LINKED ISSUE #2563: ReflectionPad2d breaks if reflection amount is too large
```python
import torch
from torch.autograd import Variable

x = Variable(torch.randn(1, 3, 2, 2), requires_grad=True)
l = torch.nn.ReflectionPad2d(5)(x).sum()
l.backward()
```
The last line segfaults.

Of course this isn't a usual situation, but it came up when trying to construct a Laplacian pyramid.

### PR DESCRIPTION
**Pull request by ssnl**
Fixes #2563 .

### CODE REVIEW

**Comment by apaszke** [comment]
I think these checks should go in to THNN/THCUNN

**Comment by SsnL** [comment]
The backends are scattered around two files basing on dim `TemporalReflectionPadding`, `SpatialReflectionPadding`. If we add 3+D kernels, there will be more. So I think that this function is a good place to add. What do you think?

**Comment by apaszke** [comment]
Eh I'm really feeling like dropping all this temporal/spatial/volumetric thing from THNN. This would make refactoring much easier. We should rethink the future of this library, because it's a terrible pain to maintain, but it needs some discussion. I'm fine with dropping them wherever. It's just usually a good idea to try to not stuff Python with too many checks, because it's super slow compared to C

**Comment by gchanan** [comment]
it's more places to change, but it will also add checks for ATen users and will be faster.

**Comment by SsnL** [comment]
Good points. I'll move these to C/CUDA.

**Comment by SsnL** [comment]
Okay! Now all the checks are in backend.

**Review by soumith — APPROVED** [review]

**Comment by fmassa** [comment]
@apaszke for the record, I think we can (should) clean up THNN/THCUNN to remove redundancies and make code maintenance easier.
